File: uploadservice/src/main/java/com/alexbbb/uploadservice/MultipartUploadTask.java
package com.alexbbb.uploadservice;

import android.content.Intent;

import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;

/**
 * Implements an HTTP Multipart upload task.
 *
 * @author alexbbb (Aleksandar Gotev)
 * @author eliasnaur
 * @author cankov
 */
public class MultipartUploadTask extends HttpUploadTask {

    private static final String NEW_LINE = "\r\n";
    private static final String TWO_HYPHENS = "--";

    private String boundary;
    private byte[] boundaryBytes;
    private byte[] trailerBytes;

    @Override
    protected void init(UploadService service, Intent intent) throws IOException {
        super.init(service, intent);
        boundary = getBoundary();
        boundaryBytes = getBoundaryBytes();
        trailerBytes = getTrailerBytes();
    }

    @Override
    protected void setupHttpUrlConnection(HttpURLConnection connection) throws IOException {
        if (params.getFiles().size() <= 1) {
            connection.setRequestProperty("Connection", "close");
        } else {
            connection.setRequestProperty("Connection", "Keep-Alive");
        }

        connection.setRequestProperty("ENCTYPE", "multipart/form-data");
        connection.setRequestProperty("Content-Type", "multipart/form-data; boundary=" + boundary);
    }

    @Override
    protected long getBodyLength() throws UnsupportedEncodingException {
        // get the content length of the entire HTTP/Multipart request body
        long parameterBytes = getRequestParametersLength();
        final long totalFileBytes = getFilesLength();

        return (parameterBytes + totalFileBytes + trailerBytes.length);
    }

    @Override
    protected void writeBody() throws IOException {
        writeRequestParameters();
        writeFiles();
        requestStream.write(trailerBytes, 0, trailerBytes.length);
    }

    private String getBoundary() {
        return "---------------------------" + System.currentTimeMillis();
    }

    private byte[] getBoundaryBytes() throws UnsupportedEncodingException {
        return (NEW_LINE + TWO_HYPHENS + boundary + NEW_LINE).getBytes("US-ASCII");
    }

    private byte[] getTrailerBytes() throws UnsupportedEncodingException {
        return (NEW_LINE + TWO_HYPHENS + boundary + TWO_HYPHENS + NEW_LINE).getBytes("US-ASCII");
    }

    private long getFilesLength() throws UnsupportedEncodingException {
        long total = 0;

        for (UploadFile file : params.getFiles()) {
            total += file.getTotalMultipartBytes(boundaryBytes.length);
        }

        return total;
    }

    private long getRequestParametersLength() throws UnsupportedEncodingException {
        long parametersBytes = 0;

        if (!params.getRequestParameters().isEmpty()) {
            for (final NameValue parameter : params.getRequestParameters()) {
                // the bytes needed for every parameter are the sum of the boundary bytes
                // and the bytes occupied by the parameter
                parametersBytes += boundaryBytes.length + parameter.getMultipartBytes().length;
            }
        }

        return parametersBytes;
    }

    private void writeRequestParameters() throws IOException {
        if (!params.getRequestParameters().isEmpty()) {
            for (final NameValue parameter : params.getRequestParameters()) {
                requestStream.write(boundaryBytes, 0, boundaryBytes.length);
                byte[] formItemBytes = parameter.getMultipartBytes();
                requestStream.write(formItemBytes, 0, formItemBytes.length);

                uploadedBodyBytes += boundaryBytes.length + formItemBytes.length;
                broadcastProgress(uploadedBodyBytes, totalBodyBytes);
            }
        }
    }

    private void writeFiles() throws IOException {
        for (UploadFile file : params.getFiles()) {
            if (!shouldContinue)
                break;

            requestStream.write(boundaryBytes, 0, boundaryBytes.length);
            byte[] headerBytes = file.getMultipartHeader();
            requestStream.write(headerBytes, 0, headerBytes.length);

            uploadedBodyBytes += boundaryBytes.length + headerBytes.length;
            broadcastProgress(uploadedBodyBytes, totalBodyBytes);

            final InputStream stream = file.getStream();
            writeStream(stream);
        }
    }

}

File: uploadservice/src/main/java/net/gotev/uploadservice/protocols/multipart/MultipartUploadTask.kt
package net.gotev.uploadservice.protocols.multipart

import net.gotev.uploadservice.BuildConfig
import net.gotev.uploadservice.HttpUploadTask
import net.gotev.uploadservice.data.NameValue
import net.gotev.uploadservice.data.UploadFile
import net.gotev.uploadservice.extensions.addHeader
import net.gotev.uploadservice.extensions.asciiByes
import net.gotev.uploadservice.extensions.utf8Bytes
import net.gotev.uploadservice.network.BodyWriter

/**
 * Implements an HTTP Multipart upload task.
 */
class MultipartUploadTask : HttpUploadTask() {

    companion object {
        private const val BOUNDARY_SIGNATURE = "-------UploadService${BuildConfig.VERSION_NAME}-"
        private const val NEW_LINE = "\r\n"
        private const val TWO_HYPHENS = "--"
    }

    private val boundary = BOUNDARY_SIGNATURE + System.nanoTime()
    private val boundaryBytes = (TWO_HYPHENS + boundary + NEW_LINE).asciiByes
    private val trailerBytes = (TWO_HYPHENS + boundary + TWO_HYPHENS + NEW_LINE).asciiByes
    private val newLineBytes = NEW_LINE.utf8Bytes

    private val NameValue.multipartHeader: ByteArray
        get() = boundaryBytes + (
            "Content-Disposition: form-data; " +
                "name=\"$name\"$NEW_LINE$NEW_LINE$value$NEW_LINE"
            ).utf8Bytes

    private val UploadFile.multipartHeader: ByteArray
        get() = boundaryBytes + (
            "Content-Disposition: form-data; " +
                "name=\"$parameterName\"; " +
                "filename=\"$remoteFileName\"$NEW_LINE" +
                "Content-Type: $contentType$NEW_LINE$NEW_LINE"
            ).utf8Bytes

    private val UploadFile.totalMultipartBytes: Long
        get() = multipartHeader.size.toLong() + handler.size(context) + newLineBytes.size.toLong()

    private fun BodyWriter.writeRequestParameters() {
        httpParams.requestParameters.forEach {
            write(it.multipartHeader)
        }
    }

    private fun BodyWriter.writeFiles() {
        for (file in params.files) {
            if (!shouldContinue) break

            write(file.multipartHeader)
            writeStream(file.handler.stream(context))
            write(newLineBytes)
        }
    }

    private val requestParametersLength: Long
        get() = httpParams.requestParameters.map { it.multipartHeader.size.toLong() }.sum()

    private val filesLength: Long
        get() = params.files.map { it.totalMultipartBytes }.sum()

    override val bodyLength: Long
        get() = requestParametersLength + filesLength + trailerBytes.size

    override fun performInitialization() {
        httpParams.requestHeaders.apply {
            addHeader("Content-Type", "multipart/form-data; boundary=$boundary")
            addHeader("Connection", if (params.files.size <= 1) "close" else "Keep-Alive")
        }
    }

    override fun onWriteRequestBody(bodyWriter: BodyWriter) {
        // reset uploaded bytes when the body is ready to be written
        // because sometimes this gets invoked when network changes
        resetUploadedBytes()
        setAllFilesHaveBeenSuccessfullyUploaded(false)

        bodyWriter.apply {
            writeRequestParameters()
            writeFiles()
            write(trailerBytes)
        }
    }
}
